### PR TITLE
Feature: add checked conversion functions to node

### LINKED ISSUE #125: Improvement: Node to subtype casting ergonomics
Casting `Node` to `Literal` is currently tedious, as the node type needs to be checked manually before doing the actual casting operation. A more convinient interface would be something like the following.

```c++
struct Node { 
    ...

    /**
      * Casts this Node into a Literal if it is one, otherwise returns the null literal.
      */ 
    Literal as_literal_checked() const noexcept;

    /**
     * Casts this Node into an IRI if it is one, otherwise returns the null iri.
     */
    IRI as_iri_checked() const noexcept;

    BNode as_bnode_checked() const noexcept;
    query::Variable as_variable_checked() const noexcept;
    ...
};
```

This addition would e.g. change this:

```c++
// given inputs
Node n1{...};
Node n2{...};

if (!n1.is_literal() || !n2.is_literal()) {
   return Literal{};
}

Literal l1 = static_cast<Literal>(n1);
Literal l2 = static_cast<Literal>(n2);

return l1 + l2;
```

```c++
// given inputs
Node n1{...};
Node n2{...};

// this works because the transformative functions on e.g. literal check for null anyways.
// by transformative functions is mean functions that take one or multiple literals and turn it into another literal
return n1.as_literal_checked() + n2.as_literal_checked();
```

### PR DESCRIPTION
**Pull request by liss-h**
Closes #125

### CODE REVIEW

File: src/rdf4cpp/rdf/writer/NQuadsWriter.cpp [hunk 1]
```diff
@@ -14,7 +14,7 @@ NQuadsWriter::operator std::string() const {
 }
 std::ostream &operator<<(std::ostream &os, const NQuadsWriter &writer) {
     for (const Quad &quad : writer.dataset_) {
-        if (not(quad.graph().is_iri() and (static_cast<IRI>(quad.graph())) == IRI::default_graph(writer.dataset_.backend().node_storage())))
+        if (!quad.graph().null() && quad.graph().as_iri() != IRI::default_graph(writer.dataset_.backend().node_storage()))
```
Comment (bigerl) [inline src/rdf4cpp/rdf/writer/NQuadsWriter.cpp:17]: NOT(a AND b) = NOT a OR NOT b -> `||`

Comment (bigerl) [inline src/rdf4cpp/rdf/writer/NQuadsWriter.cpp:17]:
```suggestion
        if (quad.graph().as_iri() != IRI::default_graph(writer.dataset_.backend().node_storage()))
```

Should be sufficient, right?

Comment (liss-h) [reply src/rdf4cpp/rdf/writer/NQuadsWriter.cpp:17]: but `null-iri != default_graph_iri` and then we would try to print a null-iri

Comment (bigerl) [reply src/rdf4cpp/rdf/writer/NQuadsWriter.cpp:17]: you are right

**Review by bigerl — APPROVED** [review]
looks good

**Pull request merged**